api/access_middleware.go

```diff
@@ -7,149 +7,104 @@
+			logger.Debug("basic auth ok")
+			ctx = auth.NewContextWithEmail(ctx, email)
+			ctx = auth.NewContextWithBasicAuthOK(ctx)
+			m.next.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
 
-	// checking basic auth in case of x-forwarded-user-email
-	basicAuthUser := m.App.Config.GetString("basicauth.username")
-	basicAuthPass := m.App.Config.GetString("basicauth.password")
-	if email, ok := basicAuthWithXForwardedUserEmail(
-		basicAuthUser, basicAuthPass, r,
-	); ok {
-		ctx := NewContextWithEmail(r.Context(), email)
-		m.next.ServeHTTP(w, r.WithContext(ctx))
-		return
-	}
+		// Basic Auth is invalid, return unauthorized
+		if result == auth.AuthenticationInvalid {
+			logger.Debug("basic auth invalid")
+			m.App.HandleError(w, http.StatusUnauthorized, "authentication failed", fmt.Errorf("invalid basic auth"))
+			return
+		}
 
-	logger := middleware.GetLogger(r.Context())
-	logger.Debug("Checking access token")
+		logger.Debug("basic auth missing")
+		// If basic auth is missing and `basicAuth.tryOauthIfUnset` is false return unauthorized
+		if !m.App.Config.GetBool("basicAuth.tryOauthIfUnset") {
+			logger.Debug("basic auth tryOauthIfUnset is false so unauthorized")
+			m.App.HandleError(w, http.StatusUnauthorized, "authentication failed", fmt.Errorf("no basic auth sent"))
+			return
+		}
+	}
 
-	accessToken := r.Header.Get("Authorization")
-	accessToken = strings.TrimPrefix(accessToken, "Bearer ")
+	if m.App.Config.GetBool("william.enabled") {
+		logger.Debug("william enabled, checking token")
+		token := r.Header.Get("Authorization")
+		if len(token) == 0 {
+			logger.Debug("token empty")
+			m.App.HandleError(w, http.StatusUnauthorized, "", errors.NewAccessError("missing access token", emptyErr))
+			return
+		}
 
-	token, err := login.GetToken(accessToken, m.App.DBClient.WithContext(r.Context()))
-	if err != nil {
-		m.App.HandleError(w, http.StatusInternalServerError, "", err)
-		return
-	}
-	if token.RefreshToken == "" {
-		m.App.HandleError(
-			w,
-			http.StatusUnauthorized,
-			"",
-			errors.NewAccessError("access token was not found on db", fmt.Errorf("access token error")),
-		)
-		return
-	}
+		token = strings.TrimPrefix(token, "Bearer ")
 
-	msg, status, err := m.App.Login.Authenticate(token, m.App.DBClient.WithContext(r.Context()))
-	if err != nil {
-		logger.WithError(err).Error("error fetching googleapis")
-		m.App.HandleError(w, http.StatusInternalServerError, "Error fetching googleapis", err)
-		return
-	}
+		if len(token) == 0 {
+			logger.Debug("no bearer token")
+			m.App.HandleError(w, http.StatusUnauthorized, "", errors.NewAccessError("Unauthorized access token", emptyErr))
+			return
+		}
 
-	if status == http.StatusBadRequest {
-		logger.WithError(err).Error("error validating access token")
-		err := errors.NewAccessError("Unauthorized access token", fmt.Errorf(msg))
-		m.App.HandleError(w, http.StatusUnauthorized, "Unauthorized access token", err)
-		return
-	}
+		logger.Debug("authenticated with william")
```

Comment:
Isnt this log a bit misleading? You're not checking with William if the token is valid or not.

Reply:
Yeah, you are right.